activesupport/lib/active_support/syntax_error_proxy.rb

```diff
@@ -43,7 +47,20 @@ def backtrace_locations
 
     private
       def parse_message_for_trace
-        __getobj__.to_s.split("\n")
+        if error_originated_in_eval_block?
+          # If the exception is coming from a call to eval, we need to keep
+          # the path of the file in which eval was called to ensure we can
+          # return the right source fragment to show the location of the
+          # error
+          location = __getobj__.backtrace_locations[0]
+          ["#{location.path}:#{location.lineno}: #{__getobj__}"]
+        else
+          __getobj__.to_s.split("\n")
+        end
+      end
+
+      def error_originated_in_eval_block?
+        __getobj__.to_s.match?(/^\(eval\)/)
```

Comment:
```suggestion
        __getobj__.to_s.start_with?("(eval")
```

This won't match anymore in 3.3: https://bugs.ruby-lang.org/issues/19755